File: Javascript/2.Settembre/28.RomanNumeralParser-7Settembre.js
//todo Data una stringa che rappresenta un numero romano, restituisci il suo valore intero.

//! I numeri romani sono costituiti dai seguenti simboli e valori:

//! Simbolo Valore
//! I 1
//! V 5
//! X 10
//! L 50
//! C 100
//! D 500
//! M 1000
//! I numeri si leggono da sinistra a destra. Se un numero più piccolo precede uno più grande, il valore viene sottratto. Altrimenti, i valori vengono sommati.

//? Tempo di esecuzione:
//* Quasi quasi finito
//?Codice scritto senza utilizzo di ChatGpt

function parseRomanNumeral(numeral) {

  let I = 1;
  let V = 5;
  let X = 10;
  let L = 50;
  let C = 100;
  let D = 500;
  let M = 1000;

  let scomposto = []
  let tot = 0

  for(let i = 0; i < numeral.length; i++) {
    if(numeral[i] === "I") scomposto.push(I)
    if(numeral[i] === "V") scomposto.push(V)
    if(numeral[i] === "X") scomposto.push(X)
    if(numeral[i] === "L") scomposto.push(L)
    if(numeral[i] === "C") scomposto.push(C)
    if(numeral[i] === "D") scomposto.push(D)
    if(numeral[i] === "M") scomposto.push(M)
  } 

  for(let j = 0; j < scomposto.length - 1; j++) {
    if(scomposto[j] < scomposto[j+1]) tot += scomposto[j+1] - scomposto[j]
    if(scomposto[j] >= scomposto[j+1]) tot += scomposto[j];
    if(scomposto[j] === scomposto[scomposto.length] && scomposto[j] >= scomposto[j-1]) tot += scomposto[j]
    if(scomposto[j] === scomposto[scomposto.length] && scomposto[j] < scomposto[j-1]) tot -= scomposto[j-1] - scomposto[j]
  } 
  return tot;
}

console.log(parseRomanNumeral("XXVI"))

//?Codice scritto da ChatGpt

function parseRomanNumeral(numeral) {
  const romanMap = {I:1, V:5, X:10, L:50, C:100, D:500, M:1000};
  let total = 0;

  for (let i = 0; i < numeral.length; i++) {
    const current = romanMap[numeral[i]];
    const next = romanMap[numeral[i+1]];

    if (next && current < next) {
      total += next - current;
      i++; // salta il prossimo perché già considerato
    } else {
      total += current;
    }
  }

  return total;
}

console.log(parseRomanNumeral("XXVI")); // 26
console.log(parseRomanNumeral("IV"));   // 4
console.log(parseRomanNumeral("MCMXC")); // 1990

//* Punti forti

//* Buona idea di mappare i simboli romani ai numeri

//* Concetto giusto: prima trasformi in numeri, poi li sommi / sottrai secondo la regola

//* Uso di scomposto come array intermedio è logico

//! Debolezze / Errori

//! Ultimo numero non viene mai aggiunto correttamente

//! Il tuo ciclo for(let j = 0; j < scomposto.length - 1; j++) salta l’ultimo elemento

//! Le condizioni con scomposto[scomposto.length] sono sbagliate: in JS l’ultimo indice è scomposto.length - 1, quindi stai leggendo undefined.

//! Condizioni troppo complesse → rischiano di sbagliare con numeri come IV o IX

//! Non gestisci errori o simboli invalidi (facoltativo, ma utile)

//todo Considerazioni: 

//*Sono andato spedito fino all'ultimo numero, poi il vuoto totale

//! N.B.

File: Python/2.Settembre/28.Roman_numeral_parser-7Settembre.py
# # //Prompt ChatGpt: questo è il modello. Invierò consegna esercizio, codice e vorrei che tu mi dessi consigli

# # //todo Consegna dell'esercizio

# Data una stringa che rappresenta un numero romano, restituisci il suo valore intero.

# I numeri romani sono composti dai simboli:
# I = 1, V = 5, X = 10, L = 50, C = 100, D = 500, M = 1000

# Regola:
# - Se un numero più piccolo precede uno più grande, viene sottratto
# - Altrimenti viene sommato

# # //todo

# Bisogna convertire ogni simbolo romano in numero e poi applicare la regola di somma/sottrazione.

# # //*Codice scritto senza utilizzo di ChatGpt

# ```python
def parse_roman_numeral(numeral):

    Roman_number = [1,5,10,50,100,500,1000]

    list_roman = []
    tot = 0;

    for roman in numeral:

        match roman:
            case "I":
                list_roman.append(Roman_number[0])
            case "V":
                list_roman.append(Roman_number[1])
            case "X":
                list_roman.append(Roman_number[2])
            case "L":
                list_roman.append(Roman_number[3])
            case "C":
                list_roman.append(Roman_number[4])
            case "D":
                list_roman.append(Roman_number[5])
            case "M":
                list_roman.append(Roman_number[6])

    indice = 1;

    for i in range(len(list_roman)):
        if i + 1 < len(list_roman) and list_roman[i] < list_roman[i+1]:
            tot -= list_roman[i]
        else:
            tot += list_roman[i]

    return tot;

print(parse_roman_numeral("IV"))
# ```

# # //*

# # //* Codice scritto da ChatGpt

# ```python
def parse_roman_numeral(numeral):
    values = {
        "I": 1,
        "V": 5,
        "X": 10,
        "L": 50,
        "C": 100,
        "D": 500,
        "M": 1000
    }

    total = 0
    prev = 0

    for char in reversed(numeral):
        value = values[char]

        if value < prev:
            total -= value
        else:
            total += value

        prev = value

    return total
# ```

# # //*

# # //* Pro e contro del mio codice

# # //* Pro

# - Hai correttamente implementato la logica di sottrazione nei casi tipo "IV".
# - Hai separato chiaramente conversione e calcolo.
# - L’idea di trasformare prima in lista numerica è valida.
# - Il ragionamento generale è corretto.

# # //! Contro

# - Usi una lista `Roman_number` con indici: soluzione poco leggibile e fragile.
# - `match` è lungo e ripetitivo (poco scalabile).
# - Variabile `indice` inutile (non viene mai usata).
# - Secondo ciclo corretto ma non ottimizzato (due passaggi invece di uno).
# - Il codice può diventare lento e pesante su stringhe lunghe.
# - Stile: `Roman_number` dovrebbe essere `snake_case` (`roman_numbers`).

# # //*

# # //* Consigli

# - Usa un dizionario invece di una lista + match:

# ```python
# values = {"I": 1, "V": 5, ...}
# ```

# - Puoi evitare completamente la lista intermedia `list_roman`.
# - Approccio più pulito: iterare da destra a sinistra tenendo traccia del valore precedente.
# - Rimuovi variabili inutili come `indice`.
# - Cerca di ridurre il problema a **un solo ciclo**.
# - Pensa sempre: “posso evitare una struttura dati intermedia?”

# # //*

# # //* Valutazione

# **7.5/10**

# Ottimo ragionamento sulla regola di sottrazione (questa è la parte più importante dell’esercizio).

# Da migliorare soprattutto:
# - pulizia del codice
# - efficienza (evitare doppio passaggio)
# - semplificazione della struttura dati

# Se lo riscrivi con un dizionario e un solo ciclo, diventa da livello junior solido.
# # //*